server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -129,48 +144,70 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
             }
             indexResponses.putIfAbsent(resp.getIndexName(), resp);
         };
-        final FailureCollector indexFailures = new FailureCollector();
-        // One for each cluster including the local cluster
-        final CountDown completionCounter = new CountDown(1 + remoteClusterIndices.size());
-        final Runnable countDown = createResponseMerger(request, fieldCapTask, completionCounter, indexResponses, indexFailures, listener);
-        final RequestDispatcher requestDispatcher = new RequestDispatcher(
-            clusterService,
-            transportService,
-            task,
-            request,
-            localIndices,
-            nowInMillis,
-            concreteIndices,
-            threadPool.executor(ThreadPool.Names.SEARCH_COORDINATION),
-            handleIndexResponse,
-            indexFailures::collect,
-            countDown
-        );
-        requestDispatcher.execute();
+        final BiConsumer<String, Exception> handleIndexFailure = (index, error) -> {
+            if (fieldCapTask.isCancelled()) {
+                releaseResourcesOnCancel.run();
+                return;
+            }
+            indexFailures.collect(index, error);
```

Comment:
I think this is still slightly racy - we should check for cancellation _after_ collecting the response to see if the newly-added response should now be dropped.

(it's ok to check first too ofc)

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/93716/commits/6c0cc91f371e24af6c63b38a5de60c2c7e5a9bb1